source/common/stats/stats_impl.h

```diff
@@ -370,12 +373,44 @@ class HistogramImpl : public Histogram, public MetricImpl {
 };
 
 /**
- * Implementation of RawStatDataAllocator that just allocates a new structure in memory and returns
- * it.
+ * Implementation of RawStatDataAllocator which accepts a block of (potentially shared) memory.
+ */
+class BlockRawStatDataAllocator : public RawStatDataAllocator {
+public:
+  /**
+   * Creates a RawStatDataAllocator with an underlying MemoryHashMap to track stat names. Optionally
+   * works with shared memory, as in HotRestartImpl. Not guaranteed to be threadsafe unless the
+   * optional stat_lock is passed in.
+   * @param options initialization parameters for MemoryHashSet
+   * @param init true if the shared-memory should be initialized on construction. See the
+   * MemoryHashSet constructor.
+   * @param memory the memory buffer for the set data.
+   * @param stat_lock an optional lock for thread safety. If nullptr, no locks are acquired.
+   * @return BlockRawStatDataAllocator newly constructed BlockRawStatDataAllocator.
+   */
+  BlockRawStatDataAllocator(const MemoryHashSetOptions& options, bool init, uint8_t* memory,
+                            Thread::BasicLockable* stat_lock)
```

Comment:
Just a note: if we decide that we want this in all cases, we might consider making this a reference.  Or if we want this class to own it, we could pass in a unique_ptr and check that it's not nullptr.